File: generator/src/java/darkness/generator/api/effects/Aurora.java
package darkness.generator.api.effects;

import darkness.generator.api.BulbSet;
import darkness.generator.api.BulbGroup;
import darkness.generator.api.BulbRGB;

import java.awt.Color;
import java.util.Arrays;
import java.util.Set;
import java.util.HashSet;
import java.util.Random;

public class Aurora extends EffectBase {
    private final BulbGroup bulbGroup;
    private final Color color;
    private final int time; // in seconds
    private final int fade; // in frames
    private final int nChangeBulbs; // number of bulbs that possibly change between blocks
    private final float minBrightness; 
    private final Random rnd;

    public Aurora(BulbGroup bulbGroup, Color color, int time, int fade, int nChangeBulbs, float minBrightness) {
        this.bulbGroup = bulbGroup;
        this.color = color;
	this.time = time;
	this.fade = fade;
	this.nChangeBulbs = nChangeBulbs;
	this.minBrightness = minBrightness;
	this.rnd = new Random(1337);
    }

    @Override
    public void run() {
	    float[] hsb_vals = Color.RGBtoHSB(color.getRed(), color.getGreen(), color.getBlue(), null);

	    int nRepeats = time * 20 / fade;

	    for (int j = 0; j < nRepeats; j++) {
		    for (int i = 0; i < nChangeBulbs; i++) {
			    int nextBulbIdx = rnd.nextInt(bulbGroup.numBulbs);
			    BulbRGB nextBulb = bulbGroup.getBulb(nextBulbIdx);
			    float nextBrightness = rnd.nextFloat();
			    while (nextBrightness <= minBrightness) {
				    nextBrightness = rnd.nextFloat();
			    }
			    Color c = Color.getHSBColor(hsb_vals[0], hsb_vals[1], nextBrightness);
			    rgbFade(nextBulb, c, fade);
		    }
		    skip(fade);
	    }
    }

    @Override
    public String toString() {
        return "Effect Aurora";
    }
}

File: generator/src/kotlin/darkness/generator/api/effects/Aurora.kt
package darkness.generator.api.effects

import darkness.generator.api.BulbGroup

import java.awt.Color
import java.util.Random

class Aurora(
    private val bulbGroup: BulbGroup,
    private val color: Color,
    /** In seconds */
    private val time: Int,
    /** In frames */
    private val fade: Int,
    /** Number of bulbs that possibly change between blocks */
    private val nChangeBulbs: Int,
    private val minBrightness: Float
) : EffectBase() {
    private val rnd: Random = Random(1337)

    override suspend fun run() {
        val (hue, saturation, _) = Color.RGBtoHSB(color.red, color.green, color.blue, null)
        val nRepeats = time * 20 / fade
        val previousTargetBrightness = mutableMapOf<Int, Float>()

        for (j in 0 until nRepeats) {
            val bulbIndexesUsedThisRound = mutableSetOf<Int>()
            for (i in 0 until nChangeBulbs) {
                val bulbIndex = rnd.nextInt(bulbGroup.numBulbs)
                val bulb = bulbGroup.getBulb(bulbIndex)
                var nextBrightness = rnd.nextFloat()
                while (nextBrightness <= minBrightness) {
                    nextBrightness = rnd.nextFloat()
                }
                // Don't skip until here, so that the random generator stays in sync with the old version
                if (!bulbIndexesUsedThisRound.add(bulbIndex)) {
                    continue
                }
                val previousColor = Color.getHSBColor(hue, saturation, previousTargetBrightness.getOrDefault(bulbIndex, 0f))
                val nextColor = Color.getHSBColor(hue, saturation, nextBrightness)
                rgbFade(bulb, previousColor, nextColor, fade)
                previousTargetBrightness[bulbIndex] = nextBrightness
            }
            skip(fade)
        }
    }

    override fun toString() = "Effect Aurora"
}
